---
title: "Investigating RNA-Seq simulations"
output:
  html_document:
    toc: true
---

```{r}
library("data.table")
library("dplyr")
library("reshape2")
library("ggplot2")
library("rjson")
```

# Loading the results

Let's first get info on all the data

```{r,echo=FALSE}
base_dir <- "../../NA12716_7/rsem/sim/30000000"
```

Helper function for cufflinks since it doesn't output effective length:

```{r}
get_oracle_mean_fl <- function(oracle) {
  tab <- table(oracle$length - oracle$eff_length)
  as.numeric(names(which(tab == max(tab)))[1])
}
```

Convenience functions for loading different datasets
```{r}
read_xprs <- function(fname) {
    xprs <- fread(fname, header = TRUE, stringsAsFactors = FALSE,
        data.table = TRUE)
    xprs
}

read_sailfish <- function(fname) {
    sf <- fread(fname, header = FALSE, skip = 5, stringsAsFactors = FALSE,
        data.table = FALSE)
    colnames(sf) <- c("target_id", "length", "tpm", "rpkm", "kpkm", "EstimatedNumKmers", "est_counts")
    sf %>%
      select(target_id, tpm, est_counts)
}

tpm_to_alpha <- function(tpm, eff_len) {
  stopifnot(length(tpm) == length(eff_len))

  alpha <- (tpm * eff_len) / 1e6
  alpha <- alpha / sum(alpha)

  alpha
}

read_rsem <- function(fname) {
  read.table(fname, header = TRUE, stringsAsFactors = FALSE) %>%
    rename(target_id = transcript_id,
      est_counts = expected_count,
      tpm = TPM, eff_len = effective_length) %>%
    select(-gene_id, -FPKM, -IsoPct)
}

read_cufflinks <- function(fname, mean_frag_len) {
  data <- data.table::fread(fname, header = TRUE)

  data <- data %>%
    mutate(tpm = FPKM * 1e6/ sum(FPKM),
      est_counts = tpm_to_alpha(tpm, length - mean_frag_len))

  data %>%
    select(target_id = tracking_id, tpm, est_counts)
}

counts_to_tpm <- function(est_counts, eff_len) {
  stopifnot( length(eff_len) == length(est_counts) )

  which_valid <- which(eff_len > 0)

  num <- (est_counts / eff_len)
  num[-which_valid] <- 0
  denom <- sum(num)

  (1e6 * num) / denom
}

read_kallisto_h5 <- function(fname, read_bootstrap = TRUE) {
  stopifnot(is(fname, "character"))

  fname <- path.expand(fname)

  if (!file.exists(fname)) {
    stop(paste0("Can't file file: '", fname, "'"))
  }

  target_id <- as.character(rhdf5::h5read(fname, "aux/ids"))
  abund <- data.frame(target_id = target_id, stringsAsFactors = FALSE)
  abund$est_counts <- as.numeric(rhdf5::h5read(fname, "est_counts"))
  abund$eff_len <- as.numeric(rhdf5::h5read(fname, "aux/eff_lengths"))
  abund$len <- as.numeric(rhdf5::h5read(fname, "aux/lengths"))

  bs_samples <- list()
  if (read_bootstrap) {
    num_bootstrap <- as.integer(rhdf5::h5read(fname, "aux/num_bootstrap"))
    if (num_bootstrap > 0) {
      cat("Found ", num_bootstrap, " bootstrap samples\n")
      bs_samples <- lapply(0:(num_bootstrap[1]-1), function(i)
        {
          .read_bootstrap_hdf5(fname, i, abund)
        })
    } else {
      cat("No bootstrap samples found\n ")
    }
  }

  abund$tpm <- counts_to_tpm(abund$est_counts, abund$eff_len)

  invisible(structure(
      list(abundance = abund,
        bootstrap = bs_samples),
      class = "kallisto"))
}

```

Load all data:

```{r,warning=FALSE,cache=TRUE,echo=FALSE}
all_oracle <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, paste0(id, ".sim.isoforms.results"))
    read.table(fname, header = TRUE, stringsAsFactors = FALSE) %>%
      select(target_id = transcript_id, counts = count, tpm = TPM,
        eff_length = effective_length, length)
  })

oracle_mfl <- lapply(all_oracle, get_oracle_mean_fl)
```
```{r,warning=FALSE,cache=TRUE}
  all_xprs <- lapply(1:20,
    function(id)
    {
      fname <- file.path(base_dir, id, "express", "results.xprs")
      read_xprs(fname)
    })

get_tot_counts <- function(xprs_res) {
  ret <- xprs_res %>%
    mutate(rho = fpkm / sum(fpkm)) %>%
    mutate(alpha = rho * eff_length  / sum(rho * eff_length)) %>%  
    filter(uniq_counts == tot_counts, tot_counts > 10) %>%
    arrange(desc(tot_counts)) %>%
    mutate(N = uniq_counts / alpha) %>%
    summarise(N = median(N)) %>%
    as.data.frame()
  ret[1,1]
}

xprs_tot_counts <- lapply(all_xprs, get_tot_counts)  



# merge results
merge_results <- function(exp_list, exp_labels, oracle) {
  stopifnot( is(exp_list, "list") )
  stopifnot( length(exp_list) == length(exp_labels) )

  exp_list <- lapply(seq_along(exp_list),
    function(i)
    {
      res <- exp_list[[i]] %>%
        select(target_id, tpm, est_counts) %>%
        data.table::data.table()
      data.table::setnames(res, "tpm", paste0("tpm_", exp_labels[i]))
      data.table::setnames(res, "est_counts", paste0("est_counts_", exp_labels[i]))
    })

  oracle <- oracle %>%
    rename(tpm_oracle = tpm, est_counts_oracle = counts)

  all_res <- Reduce(function(x, y) inner_join(x, y, by = c("target_id")), exp_list)


  melt_by <- function(data, unit_by) {
    m_unit <- data %>%
      select(target_id, starts_with(unit_by)) %>%
      reshape2::melt(id.vars = "target_id", variable.name = "method")
    ret <- data.table::data.table(oracle) %>%
      select(target_id, starts_with(unit_by)) %>%
      inner_join(data.table::data.table(m_unit), by = "target_id") %>%
      rename(estimate = value)
    data.table::setnames(ret, paste0(unit_by, "_oracle"), "oracle")
    ret
  }

  m_tpm <- melt_by(all_res, "tpm")
  m_est_counts <- melt_by(all_res, "est_counts")

  all_res <- all_res %>%
    inner_join(data.table::data.table(oracle), by = "target_id")

  structure(list(all_data = all_res, m_tpm = m_tpm, m_est_counts = m_est_counts),
    class = "merged_res")
}

filtered_summary <- function(mres, filter_exp) {
  stopifnot( is(mres, "merged_res") )
  do_filter <- if (missing(filter_exp)) {
    FALSE
  } else {
    filter_exp <- deparse(substitute(filter_exp))
    filtered_ids <- mres$all_data %>%
      filter_(.dots = list(filter_exp)) %>%
      select(target_id)
    TRUE
  }

  both_res <- lapply(list(mres$m_tpm, mres$m_est_counts),
    function(res)
    {
      if (do_filter) {
        res <- data.table(res) %>%
          inner_join(data.table(filtered_ids), by = c("target_id"))
      }

      res %>%
        group_by(method) %>%
        summarise(
          pearson = cor(estimate, oracle, method = "pearson"),
          spearman = cor(estimate, oracle, method = "spearman"),
          med_scaled_err = median(abs(scaled_error(estimate, oracle)),
              na.rm = TRUE),
          med_per_err = median(abs(percent_error(estimate, oracle)))
          )
    })

  setNames(both_res, c("tpm", "est_counts"))
}

scaled_error <- function(estimate, truth) {
  2 *(estimate - truth)  / (estimate + truth)
}

percent_error <- function(estimate, truth) {
  (estimate - truth) / truth
}

```

```{r,warning=FALSE,cache=TRUE}
all_cufflinks <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "cufflinks", "isoforms.fpkm_tracking")
    read_cufflinks(fname, oracle_mfl[[id]]) %>%
      mutate(est_counts = est_counts * xprs_tot_counts[[id]])
  })
```

```{r,warning=FALSE,cache=TRUE}
all_rsem <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "rsem", "out.isoforms.results")
    read_rsem(fname)
  })
```
```{r,warning=FALSE}
all_sailfish <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "sailfish", "quant.sf")
    x <- read_sailfish(fname)
    x
  })
```

Sailfish counts look odd:
```{r}
lapply(all_sailfish, function(x) sum(x$est_counts)) %>%
  unlist()
```

Let's correct them with the true total counts:

```{r}
all_sailfish <- lapply(all_sailfish, function(x) {
  #x$est_counts <- x$est_counts * (30e6/sum(x$est_counts))
  # each end of read is processed independently
  x$est_counts <- x$est_counts  / 2
  
  x
})  
```

```{r,warning=FALSE}
all_kallisto <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "kallisto_soft", "abundance.h5")
    read_kallisto_h5(fname, FALSE)$abundance
  })
```


```{r}
mr <- lapply(seq_along(all_oracle),
  function(id)
  {
    merge_results(
      list(all_sailfish[[id]], all_xprs[[id]], all_cufflinks[[id]], all_rsem[[id]],
           all_kallisto[[id]]),
      c("sailfish", "express", "cufflinks", "rsem", "kallisto"),
      all_oracle[[id]])
  })
```

```{r}
no_filt_all <- lapply(mr,
    function(res)
    {
        filtered_summary(res)$est_counts
    }) %>%
    rbind_all()
```


```{r,fig.width=14,fig.height=10}
no_filt_all %>%
  mutate(method = sub("est_counts_", "", method)) %>%
  mutate(method = factor(method, levels = c("cufflinks", "sailfish", "express", "kallisto", "rsem"))) %>%
#   mutate(method = factor(method, labels = c(
#                                             "TopHat2\n+\nCufflinks",
#                                             "Sailfish",
#                                             "Bowtie2\n+\neXpress",
#                                             "kallisto",
#                                             "Bowtie2\n+\nRSEM"
#                                             ))) %>% 
  group_by(method) %>%
  summarise(med_rel_diff = mean(med_scaled_err)) %>%
  ggplot(aes(method, med_rel_diff)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(med_rel_diff, 2),
                  y = med_rel_diff + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, .6)
```

```{r}
no_filt_tpm <- lapply(mr,
    function(res)
    {
        filtered_summary(res)$tpm
    }) %>%
    rbind_all()
```

Looking at TPM:

```{r,fig.width=14,fig.height=10}
no_filt_tpm %>%
  mutate(method = sub("tpm_", "", method)) %>%
  group_by(method) %>%
  summarise(med_rel_diff = mean(med_scaled_err)) %>%
  ggplot(aes(method, med_rel_diff)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(med_rel_diff, 2),
                  y = med_rel_diff + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 0.6)
```


## Filtering

```{r}
filt_tpm_1_all <- lapply(mr,
    function(res)
    {
        filtered_summary(res, tpm_oracle >= 1.0)$est_counts
    }) %>%
    rbind_all()
```



# Timings

```{r,echo=FALSE}
time_base_dir <- "../../benchmarks/NA12716_7/rsem/sim/30000000/"
```

```{r}
mr_upd <- lapply(mr,
       function(res) {
         res$all_data <- filter(res$all_data, length >= 500)
         res$m_est_counts <- inner_join(
           dplyr::select(res$all_data, target_id),
           res$m_est_counts,
           by = "target_id")
          res$m_tpm <- inner_join(
           dplyr::select(res$all_data, target_id),
           res$m_tpm,
           by = "target_id")
         
         res$m_tpm %>%
           group_by(method) %>%
           mutate(oracle = oracle / sum(oracle), estimate / sum(estimate))
         res
       })
```


```{r}
filt_length <- lapply(mr_upd,
    function(res)
    {
        filtered_summary(res, tpm_oracle >= 1)$tpm
    }) %>%
    rbind_all()
```

```{r}
filt_length %>%
  mutate(method = sub("tpm_", "", method)) %>%
  group_by(method) %>%
  summarise(med_rel_diff = mean(med_scaled_err)) %>%
  ggplot(aes(method, med_rel_diff)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(med_rel_diff, 2),
                  y = med_rel_diff + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)
```



```{r,echo=FALSE}
load_sec <- function(path) {
  timings <- fromJSON(file = path)
  data.frame(seconds = timings$wall_clock_times$s) %>%
    mutate(minutes = seconds / 60.0)
}

load_timings <- function(nsim, method_name, fname) {
  lapply(1:nsim,
    function(id)
    {
      load_sec(file.path(time_base_dir, id, fname)) %>%
        mutate(method = method_name, id = id)
    }) %>%
    rbind_all()  
}
```

```{r,echo=FALSE}
sailfish_times <- load_timings(20, "sailfish", "sailfish.json")
tophat_times <- load_timings(20, "tophat", "tophat.json")
cufflinks_times <- load_timings(20, "cufflinks", "cufflinks.json")
bwt2_times <- load_timings(20, "bwt2", "bwt2.json")
express_times <- load_timings(20, "express", "express.json")
bwt2_rsem_times <- load_timings(20, "bwt2_rsem", "bwt2_rsem.json")
rsem_times <- load_timings(20, "rsem", "rsem.json")
kallisto_times <- load_timings(20, "kallisto", "kallisto.json")
#salmon_times <- load_timings(20, "salmon", "salmon.json")
```

Merge to get pipeline results

```{r}
total_time_express <- sum(bwt2_times$seconds) + max(express_times$seconds)
total_time_sailfish <- sum(sailfish_times$seconds)
total_time_cufflinks <- sum(tophat_times$seconds) + sum(cufflinks_times$seconds)
total_time_rsem <- sum(bwt2_rsem_times$seconds) + sum(rsem_times$seconds)
total_time_kallisto <- max(kallisto_times$seconds)

total_time_summary <- data.frame(
  method = factor(c("Bowtie2\n+\neXpress", "Sailfish",
                    "TopHat2\n+\nCufflinks", "Bowtie2\n+\nRSEM",
                    "kallisto"),
                  levels = c("kallisto", "Sailfish", "Bowtie2\n+\neXpress",
                             "Bowtie2\n+\nRSEM", "TopHat2\n+\nCufflinks")),
  seconds = c(total_time_express, total_time_sailfish, total_time_cufflinks,
              total_time_rsem, total_time_kallisto),
  stringsAsFactors = FALSE) %>%
  mutate(minutes = seconds / 60) %>%
  arrange(seconds)
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2",
  "#D55E00", "#CC79A7", "#000000")
```

```{r}
total_time_summary$method <- reorder(total_time_summary$method, -total_time_summary$minutes)
```



```{r}
total_time_summary <- total_time_summary %>%
  mutate(pos = minutes)
total_time_summary %>%
  ggplot(aes(method, minutes)) +
    geom_bar(stat="identity", fill = "dark grey") +
    geom_text(aes(label = round(minutes, 2), y = pos),
              colour = "black", size = 7) +
    xlab("Method") +
    ylab("Minutes") +
    theme_bw(25) +
    theme(legend.position = "none")
```

# Paralog analysis

```{r}
dgd <- fread("../dgd_Hsa_all_v.tsv", header=TRUE)
dgd_unique <- dgd %>%
  dplyr::select(gene_id = ENS_ID) %>%
  distinct()
```
```{r}
library(biomaRt)
mart <- useMart(biomart = "ensembl", dataset
      = "hsapiens_gene_ensembl")
gene_names <- getBM(attributes
      = c("ensembl_transcript_id","ensembl_gene_id"), mart
      = mart)
```

```{r}
gene_names <- gene_names %>%
  rename(target_id = ensembl_transcript_id, gene_id = ensembl_gene_id) %>%
  data.table()
dgd_unique <- dgd_unique %>%
  data.table() %>%
  inner_join(gene_names, by = "gene_id")
mr_paralog <- lapply(mr,
                     function(x)
                       {
                        x$m_est_counts <- inner_join(x$m_est_counts, dgd_unique, by = "target_id")
                        x
                       })
```

sanity check:

```{r}
n_paralog_trans <- dplyr::select(mr_paralog[[1]]$m_est_counts, target_id) %>%
  distinct() %>%
  nrow()
n_paralog_trans
```

```{r}
# nf - no filter
nf_paralog <- lapply(mr_paralog,
    function(res)
    {
        filtered_summary(res)$est_counts
    }) %>%
    rbind_all()
```

```{r}

nf_paralog %>%
  mutate(method = sub("est_counts_", "", method)) %>%
  mutate(method = factor(method, levels = c("cufflinks", "sailfish", "express", "kallisto", "rsem"))) %>%
  mutate(method = factor(method, labels = c(
                                            "TopHat2\n+\nCufflinks",
                                            "Sailfish",
                                            "Bowtie2\n+\neXpress",
                                            "kallisto",
                                            "Bowtie2\n+\nRSEM"
                                            ))) %>% 
  group_by(method) %>%
  summarise(med_rel_diff = mean(med_scaled_err)) %>%
#   filter(!grepl("cufflinks", method)) %>%
  ggplot(aes(method, med_rel_diff)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(med_rel_diff, 2),
                  y = med_rel_diff + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)
```

```{r}
which_paralogs <- mr_paralog[[1]]$m_est_counts %>%
  dplyr::select(target_id) %>%
  distinct() %>%
  mutate(paralog = TRUE)

counts_dist <- mr[[1]]$m_est_counts %>%
  dplyr::select(target_id, oracle) %>%
  distinct() %>%
  left_join(which_paralogs, by = c("target_id"))

counts_dist <- counts_dist %>%
  mutate(paralog = ifelse(is.na(paralog), FALSE, TRUE)) %>%
  mutate(lab = ifelse(paralog, "Duplicated Genes Database", "Remaining transcripts")) %>%
  mutate(lab = factor(lab))
```

```{r}
ggplot(counts_dist, aes(oracle)) +
  geom_histogram(aes(y = ..density..)) +
  scale_x_log10() +
  facet_wrap(~ lab) +
  theme_bw(20) +
  xlab("true counts")
```

```{r}
hi <- dplyr::select(all_kallisto[[1]], target_id, eff_len) %>%
  inner_join(dplyr::select(all_rsem[[1]], target_id, rsem_eff_len = eff_len), by = 'target_id')

ggplot(hi, aes(eff_len, rsem_eff_len)) +
  geom_point() +
  xlim(0, 250) +
  ylim(0, 250)
```