---
title: "Investigating RNA-Seq simulations"
output:
  html_document:
    toc: true
---

```{r}
library("data.table")
library("dplyr")
library("reshape2")
library("sleuth")
library("ggplot2")
#library("rjson")
library('jsonlite')
library('mamabear')
```

# Loading the results

Let's first get info on all the data

```{r,echo=FALSE}
base_dir <- "~/kallisto_paper_analysis/simulations_hisat/NA12716_7/rsem/sim/30000000"
```

Helper function for cufflinks since it doesn't output effective length:

```{r}
get_oracle_mean_fl <- function(oracle) {
  tab <- table(oracle$length - oracle$eff_length)
  as.numeric(names(which(tab == max(tab)))[1])
}
```

```{r}
read_salmon_0.4.2 <- function(fname) {
  res <- read.table(fname, stringsAsFactors = FALSE, header = FALSE)
  colnames(res) <- c("target_id", "length", "tpm", "est_counts")
  
  res
}
```

Load all data:

```{r,warning=FALSE,cache=TRUE,echo=FALSE}
all_oracle <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, paste0(id, ".sim.isoforms.results"))
    result <- read.table(fname, header = TRUE, stringsAsFactors = FALSE) %>%
      select(target_id = transcript_id, counts = count, tpm = TPM,
        eff_length = effective_length, length)
    # filter(result, target_id  != 'ENST00000536425')
    result
  })

#oracle_mfl <- lapply(all_oracle, get_oracle_mean_fl)
```
```{r,warning=FALSE,cache=TRUE}
all_xprs <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "express", "results.xprs")
    read_xprs(fname)
  })

get_tot_counts <- function(xprs_res) {
  ret <- xprs_res %>%
    mutate(rho = fpkm / sum(fpkm)) %>%
    mutate(alpha = rho * eff_length  / sum(rho * eff_length)) %>%  
    filter(uniq_counts == tot_counts, tot_counts > 10) %>%
    arrange(desc(tot_counts)) %>%
    mutate(N = uniq_counts / alpha) %>%
    summarise(N = median(N)) %>%
    as.data.frame()
  ret[1,1]
}

xprs_tot_counts <- lapply(all_xprs, get_tot_counts)  
```

```{r,warning=FALSE,cache=TRUE}
all_cufflinks_tophat <- lapply(1:20,
  function(id) {
    fname <- file.path(base_dir, id, "cufflinks/tophat", "isoforms.fpkm_tracking")
    
    cufflinks <- read_cufflinks(fname, oracle_mfl[[id]])
    cufflinks <- inner_join(cufflinks,
      select(all_oracle[[id]], target_id, eff_length),
      by = 'target_id')
    stopifnot(sum(is.na(cufflinks$eff_length)) == 0)
    cufflinks <- mutate(cufflinks, est_counts = tpm_to_alpha(tpm, eff_length))
    cufflinks <- mutate(cufflinks, est_counts = est_counts * xprs_tot_counts[[id]])
    
    cufflinks
  })

all_cufflinks_hisat <- lapply(1:20,
  function(id) {
    fname <- file.path(base_dir, id, "cufflinks/hisat", "isoforms.fpkm_tracking")
    
    cufflinks <- read_cufflinks(fname, oracle_mfl[[id]])
    cufflinks <- inner_join(cufflinks,
      select(all_oracle[[id]], target_id, eff_length),
      by = 'target_id')
    stopifnot(sum(is.na(cufflinks$eff_length)) == 0)
    cufflinks <- mutate(cufflinks, est_counts = tpm_to_alpha(tpm, eff_length))
    cufflinks <- mutate(cufflinks, est_counts = est_counts * xprs_tot_counts[[id]])
    
    cufflinks
  })
```

```{r}
lower_bound <- 1e-100
small_filter <- function(x) {
  ifelse(x < lower_bound, 0, x)
}
```


```{r,warning=FALSE,cache=TRUE}
all_rsem <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "rsem", "out.isoforms.results")
    read_rsem(fname)
  })
```
```{r,warning=FALSE}
all_sailfish <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "sailfish", "quant.sf")
    x <- read_sailfish(fname)
    x
  })
```

Sailfish counts look odd:
```{r}
lapply(all_sailfish, function(x) sum(x$est_counts)) %>%
  unlist()
```

Let's correct them with the true total counts:

```{r}
all_sailfish <- lapply(all_sailfish, function(x) {
  #x$est_counts <- x$est_counts * (30e6/sum(x$est_counts))
  x$est_counts <- x$est_counts  / 2
  x
})  
```

```{r,warning=FALSE}
all_kallisto <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "kallisto", "abundance.h5")
    read_kallisto_h5(fname, FALSE)$abundance
  })
```

```{r,warning=FALSE}
#all_kallisto_np <- lapply(1:20,
#  function(id)
#  {
#    fname <- file.path(base_dir, id, "kallisto_np", "abundance.h5")
#    read_kallisto_h5(fname, FALSE)$abundance
#  })
```

```{r}
all_salmon <- lapply(1:20,
  function(id)
  {
    fname <- file.path(base_dir, id, "salmon", "quant.sf")
    read_salmon_0.4.2(fname)
  })
```

```{r}
all_xprs <- lapply(all_xprs,
  function(x) {
    mutate(x, tpm = counts_to_tpm(est_counts, eff_length))
  })
```

```{r}
all_cufflinks_hisat <- lapply(all_cufflinks_hisat,
  function(x) {
    mutate(x, est_counts = small_filter(est_counts))
  })
all_cufflinks_tophat <- lapply(all_cufflinks_tophat,
  function(x) {
    mutate(x, est_counts = small_filter(est_counts))
  })
all_sailfish <- lapply(all_sailfish,
    function(x) {
    mutate(x, est_counts = small_filter(est_counts))
  })
all_xprs <- lapply(all_xprs,
  function(x) {
    mutate(x, est_counts = small_filter(est_counts))
  })
all_rsem <- lapply(all_rsem,
  function(x) {
    mutate(x, est_counts = small_filter(est_counts))
  })
all_kallisto <- lapply(all_kallisto,
  function(x) {
    mutate(x, est_counts = small_filter(est_counts))
    })
```


```{r}
mr <- lapply(seq_along(all_oracle),
  function(id)
  {
    merge_results(
      list(all_sailfish[[id]],
           all_xprs[[id]],
           all_cufflinks_tophat[[id]],
           all_cufflinks_hisat[[id]],
           all_rsem[[id]],
           all_kallisto[[id]],
#           all_kallisto_np[[id]],
           all_salmon[[id]]),
      c("Sailfish",
        "Bowtie2\n+\neXpress",
        "TopHat2\n+\nCufflinks",
        "HISAT\n+\nCufflinks",
        "Bowtie2\n+\nRSEM",
        "kallisto",
#        "kallisto (new params)",
        "Salmon\n(0.4.2)"),
      all_oracle[[id]])
  })
```

```{r}
no_filt_all <- lapply(mr,
    function(res)
    {
        filtered_summary(res)$est_counts
    }) %>%
    rbind_all()
```

```{r}
no_filt_all %>%
  group_by(method) %>%
  summarise(mean_pearson = mean(pearson),
            mean_spearman = mean(spearman),
            mean_mrd = mean(mrd)) %>%
  arrange(mean_mrd)
```

```{r}
filter_perfect <- lapply(mr,
    function(res)
    {
        filtered_summary(res)$est_counts
    }) %>%
    rbind_all() %>% 
  group_by(method) %>% 
  summarize(mean_pearson = mean(pearson),
            mean_spearman = mean(spearman),
    mean_mrd = mean(mrd)        
    ) %>%
  arrange(mean_mrd)
```


This next plot shows performance on mean of median relative difference which appears in the paper:

```{r}
filter_perfect %>% 
  filter(!grepl('mon', method)) %>% 
  mutate(method = sub('est_counts_', '', as.character(method))) %>%
  mutate(method = factor(method, arrange(., desc(mean_mrd))[['method']])) %>%
  ggplot(aes(method, mean_mrd)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(mean_mrd, 2)),
      position=position_dodge(width=0.9), vjust=-0.25,
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)
```

### Exploring some different metrics

```{r}
filtered_summary(mr[[1]], eff_length > 100)$est_counts %>% 
  select(method, mrd)
```

method 1: no filtering whatsoever

```{r}
temp <- mr[[1]]$m_est_counts

td <- temp %>% 
  group_by(method) %>% 
  mutate(diff = relative_difference(estimate, oracle, na_zeroes = FALSE,
    normalize_counts = TRUE))
td %>% 
  group_by(method) %>% 
  summarize(median(abs(diff)))
```

method 2: filter on the length

```{r}
small_transcripts <- filter(mr[[1]]$all_data, eff_length < 100)
nrow(small_transcripts)

remove_small <- anti_join(td, small_transcripts, by = 'target_id')
nrow(remove_small)

remove_small %>% 
  group_by(method) %>% 
  summarize(median(abs(diff)))  
```

method 3: filter on 0 scores for everyone

```{r}
to_remove <- td %>% 
  group_by(target_id) %>% 
  summarize(y = all(abs(diff) <= .Machine$double.eps^2)) %>% 
  filter(y)
nrow(to_remove)

no_zeros <- anti_join(td, to_remove, by = 'target_id')
no_zeros %>% 
  group_by(method) %>% 
  summarize(median(abs(diff)))
```

```{r}
no_zeros %>% 
  filter(oracle>0) %>%   
  ggplot(aes(abs(diff), color = method)) +
  stat_ecdf() +
  xlim(0, 2)
```

method four: filter on length and median relative difference 0 for everyone

```{r}
both_filters <- anti_join(no_zeros, small_transcripts, by = 'target_id')
nrow(both_filters)

both_filters %>% 
  group_by(method) %>% 
  summarize(median(abs(diff)))

to_remove <- temp %>% 
  group_by(method) %>% 
  mutate(diff = relative_difference(estimate, oracle, na_zeroes = FALSE, normalize_counts = TRUE)) %>% 
  group_by(target_id) %>% 

      
temp_to <- anti_join(temp, to_remove, by = 'target_id')
temp_3 <- anti_join(temp_to, small_transcripts, by = 'target_id')

temp_3 %>% 
  group_by(method) %>% 
  mutate(diff = relative_difference(estimate, oracle, na_zeroes = FALSE, normalize_counts = TRUE)) %>% 
  group_by(method) %>% 
  summarize(med = median(abs(diff)))
```

Note: this figure is out of date and therefore will not be evaluated

```{r,fig.width=14,fig.height=10, eval = FALSE}
tmp <- no_filt_all %>%
  mutate(method = sub("est_counts_", "", method))
tmp <- tmp %>%
  group_by(method) %>%
  summarise(med_rel_diff = mean(med_rel_diff_no_zeroes))
tmp <- tmp %>%
  mutate(method = factor(method)) %>%
  mutate(method = factor(method, levels(method)[order(-tmp$med_rel_diff)]))


levels(tmp$method) <- tmp$method

tmp %>%
  ggplot(aes(method, med_rel_diff)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(med_rel_diff, 2),
                  y = med_rel_diff + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)

#ggsave("~/MRD.pdf", width = 10, height = 5, units = "in")
```

```{r}
no_filt_tpm <- lapply(mr,
    function(res)
    {
        filtered_summary(res)$tpm
    }) %>%
    rbind_all()
```

```{r}
no_filt_tpm %>%
  group_by(method) %>%
  summarise(mean_pearson = mean(pearson),
            mean_spearman = mean(spearman),
            mean_rel_diff_no_zeroes = mean(med_rel_diff_no_zeroes)) %>%
  arrange(mean_rel_diff_no_zeroes)
```


These plots show the cufflinks is behaving strangely.

```{r}
ggplot(filter(mr[[1]]$m_tpm, grepl('TopHat', method)), aes(oracle +1, estimate +1)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10()

ggplot(filter(mr[[1]]$m_tpm, grepl('RS', method)), aes(oracle +1, estimate +1)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10()
```

Looking at TPM:

```{r,fig.width=14,fig.height=10}
no_filt_tpm %>%
  mutate(method = sub("tpm_", "", method)) %>%
  group_by(method) %>%
  #summarise(med_rel_diff = mean(med_scaled_err)) %>%
  summarise(mean_rel_diff_no_zeroes = mean(med_rel_diff_no_zeroes)) %>%  
  ggplot(aes(method, mean_rel_diff_no_zeroes)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(mean_rel_diff_no_zeroes, 2),
                  y = mean_rel_diff_no_zeroes + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)
```


## Filtering

Cufflinks seems to do poorly when the length is short. Because of this, we will filter out things with short effective length.

```{r}
filter_length <- lapply(mr,
    function(res)
    {
       filtered_summary(res, eff_length > 100)$est_counts
    }) %>%
    rbind_all()
```

```{r}
summarize_filter_length <- filter_length %>%
  group_by(method) %>%
  summarise(mean_pearson = mean(pearson),
            mean_spearman = mean(spearman),
            mean_mrd  = mean(mrd)
    ) %>%
  arrange(mean_mrd)
summarize_filter_length
```

```{r}
summarize_filter_length %>%
  mutate(method = sub('est_counts_', '', as.character(method))) %>%
  mutate(method = factor(method, arrange(., desc(mean_rel_diff_no_zeroes))[['method']])) %>%
  ggplot(aes(method, mean_rel_diff_no_zeroes)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(mean_rel_diff_no_zeroes, 2)),
      position=position_dodge(width=0.9), vjust=-0.25,
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)
```

```{r}
filt_counts_tpm_1_all <- lapply(mr,
    function(res)
    {
      filtered_summary(res, tpm_oracle >= 1.0)$est_counts
       #filtered_summary(res, eff_length > 100)$est_counts
    }) %>%
    rbind_all()
```

```{r}
filt_counts_tpm_1_all %>% 
  mutate(method = sub("est_counts_", "", method)) %>%
  group_by(method) %>%
  summarise(med_rel_diff = mean(med_scaled_err)) %>%
  ggplot(aes(method, med_rel_diff)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(med_rel_diff, 2),
                  y = med_rel_diff + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)

```

# Timings

```{r,echo=FALSE}
time_base_dir <- "~/kallisto_paper_analysis/simulations_hisat/benchmarks/NA12716_7/rsem/sim/30000000/"
```

```{r}
mr_upd <- lapply(mr,
       function(res) {
         res$all_data <- filter(res$all_data, length >= 500)
         res$m_est_counts <- inner_join(
           dplyr::select(res$all_data, target_id),
           res$m_est_counts,
           by = "target_id")
          res$m_tpm <- inner_join(
           dplyr::select(res$all_data, target_id),
           res$m_tpm,
           by = "target_id")
         
         res$m_tpm %>%
           group_by(method) %>%
           mutate(oracle = oracle / sum(oracle), estimate / sum(estimate))
         res
       })
```


```{r}
filt_length <- lapply(mr_upd,
    function(res)
    {
        filtered_summary(res, tpm_oracle >= 1)$tpm
    }) %>%
    rbind_all()
```

```{r}
filt_length %>%
  mutate(method = sub("tpm_", "", method)) %>%
  group_by(method) %>%
  summarise(med_rel_diff = mean(med_scaled_err)) %>%
  ggplot(aes(method, med_rel_diff)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(med_rel_diff, 2),
                  y = med_rel_diff + 0.06),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)
```



```{r,echo=FALSE}
load_sec <- function(fname) {
  timings <- fromJSON(fname)
  data.frame(seconds = timings$wall_clock_times$s) %>%
    mutate(minutes = seconds / 60.0)
}

load_timings <- function(nsim, method_name, fname, type) {
  lapply(1:nsim,
    function(id)
    {
      load_sec(file.path(time_base_dir, id, fname)) %>%
        mutate(method = method_name, id = id, type = type)
    }) %>%
    rbind_all()  
}
```

load the quantification times

```{r,echo=FALSE}
sailfish_times <- load_timings(20, "Sailfish", "sailfish.json", 'quant')
cufflinks_tophat_times <- load_timings(20, "TopHat2\n+\nCufflinks",
  "cufflinks_tophat.json", 'quant')
cufflinks_hisat_times <- load_timings(20, "HISAT\n+\nCufflinks",
  "cufflinks_hisat.json", 'quant')
express_times <- load_timings(20, "Bowtie2\n+\neXpress", "express.json", 'quant')
rsem_times <- load_timings(20, "Bowtie2\n+\nRSEM", "rsem.json", 'quant')
kallisto_times <- load_timings(20, "kallisto", "kallisto.json", 'quant')
salmon_times <- load_timings(20, "Salmon", "salmon.json", 'quant')
#salmon_times <- load_timings(20, "salmon", "salmon_0.4.2.json")
```

load the alignments times

```{r}
tophat_times <- load_timings(20, "TopHat2\n+\nCufflinks", "tophat.json", 'align')
hisat_times <- load_timings(20, "HISAT\n+\nCufflinks", "hisat.json", 'align')
bwt2_times <- load_timings(20, "Bowtie2\n+\neXpress", "bwt2.json", 'align')
bwt2_rsem_times <- load_timings(20, "Bowtie2\n+\nRSEM", "bwt2_rsem.json", 'align')
```

```{r}
all_times <- bind_rows(
  sailfish_times,
  cufflinks_tophat_times,
  cufflinks_hisat_times,
  express_times,
  rsem_times,
  kallisto_times,
  salmon_times,
  tophat_times,
  hisat_times,
  bwt2_times,
  bwt2_rsem_times)
```

get the total times

```{r}
serial_times <- all_times %>% 
  filter(grepl('Top|HI|RSE|fish|mon', method)) %>% 
  group_by(method) %>% 
  summarize(total = sum(minutes))
express_align <- all_times %>% 
  filter(grepl('press', method)) %>% 
  filter(type == 'align') %>% 
  group_by(method) %>% 
  summarize(total = sum(minutes))
parallel_times <- all_times %>% 
  filter(grepl('press|isto', method))  %>% 
  filter(type == 'quant') %>% 
  group_by(method) %>% 
  summarize(total = max(minutes))
total_times <- bind_rows(serial_times, express_align, parallel_times) %>% 
  group_by(method) %>% 
  summarize(total = sum(total))
```
  
```{r}
parallel_times <- parallel_times %>% 
  mutate(type = 'quant', id = 0) %>% 
  rename(minutes = total)
plot_times <- bind_rows(
  parallel_times,
  bwt2_times,
  filter(all_times, grepl('Top|HI|RSE|fish', method)))
```

```{r}
plot_times %>% 
  mutate(method = factor(method, arrange(total_times, desc(total))[['method']])) %>%
  #mutate(type = factor(type, levels = c('quant', 'align'))) %>% 
  ggplot(aes(method, minutes, fill = type, group = as.factor(id)),
    color = 'black') +
  geom_bar(color = 'black', stat = 'identity')
```

**old results**

Merge to get pipeline results

```{r}
total_time_express <- sum(bwt2_times$seconds) + max(express_times$seconds)
total_time_sailfish <- sum(sailfish_times$seconds)
total_time_cufflinks <- sum(tophat_times$seconds) + sum(cufflinks_times$seconds)
total_time_rsem <- sum(bwt2_rsem_times$seconds) + sum(rsem_times$seconds)
total_time_kallisto <- max(kallisto_times$seconds)
total_time_salmon <- sum(salmon_times$seconds)

total_time_summary <- data.frame(
  method = factor(c("Bowtie2\n+\neXpress", "Sailfish",
                    "TopHat2\n+\nCufflinks", "Bowtie2\n+\nRSEM",
                    "kallisto", "Salmon\n(0.4.2)")),
  seconds = c(total_time_express,
              total_time_sailfish,
              total_time_cufflinks,
              total_time_rsem,
              total_time_kallisto,
              total_time_salmon),
  stringsAsFactors = FALSE) %>%
  mutate(minutes = seconds / 60)

%>%
  arrange(seconds)

total_time_summary <- total_time_summary %>%
  mutate(method = factor(method, method[order(-minutes)]))

levels(tmp_time$method)

cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2",
  "#D55E00", "#CC79A7", "#000000")
```


Black version of the plot
```{r}
total_time_summary %>%
  ggplot(aes(method, minutes)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(minutes, 2),
                  y = minutes + 100),
              colour = "black", size = 5.5) +
    xlab("") + 
    theme_bw(20)

ggsave("~/times.pdf", width = 7, height = 6)
```

```{r}
total_time_summary$method <- reorder(total_time_summary$method, -total_time_summary$minutes)
```



```{r}
total_time_summary <- total_time_summary %>%
  mutate(pos = minutes)
total_time_summary %>%
  ggplot(aes(method, minutes)) +
    geom_bar(stat="identity", fill = "dark grey") +
    geom_text(aes(label = round(minutes, 2), y = pos),
              colour = "black", size = 7) +
    xlab("Method") +
    ylab("Minutes") +
    theme_bw(25) +
    theme(legend.position = "none")
```

# Paralog analysis

```{r}
dgd <- fread("../dgd_Hsa_all_v.tsv", header=TRUE)
dgd_unique <- dgd %>%
  dplyr::select(gene_id = ENS_ID) %>%
  distinct()
```
```{r}
library(biomaRt)
mart <- useMart(biomart = "ensembl", dataset
  = "hsapiens_gene_ensembl")
gene_names <- getBM(attributes
  = c("ensembl_transcript_id","ensembl_gene_id"), mart
  = mart)
```

```{r}
gene_names <- gene_names %>%
  rename(target_id = ensembl_transcript_id, gene_id = ensembl_gene_id) %>%
  data.table()
dgd_unique <- dgd_unique %>%
  data.table() %>%
  inner_join(gene_names, by = "gene_id")
mr_paralog <- lapply(mr,
  function(x)
    {
    x$m_est_counts <- inner_join(x$m_est_counts, dgd_unique, by = "target_id")
    x
    })
```

sanity check:

```{r}
n_paralog_trans <- dplyr::select(mr_paralog[[1]]$m_est_counts, target_id) %>%
  distinct() %>%
  nrow()
n_paralog_trans
```

```{r}
# nf - no filter
nf_paralog <- lapply(mr_paralog,
    function(res)
    {
        filtered_summary(res)$est_counts
    }) %>%
    rbind_all()
```

```{r}
nf_paralog %>%
  filter(!grepl('Salmon', method)) %>% 
  mutate(method = sub("est_counts_", "", method)) %>%
  group_by(method) %>%
  summarise(mean_mrd = mean(mrd)) %>%
  mutate(method = factor(method, arrange(., desc(mean_mrd))[['method']])) %>%  
#   filter(!grepl("cufflinks", method)) %>%
  ggplot(aes(method, mean_mrd)) +
    geom_bar(stat="identity", fill = "black") +
    geom_text(aes(label = round(mean_mrd, 2)),
      position=position_dodge(width=0.9), vjust=-0.25,
              colour = "black", size = 5.5) +  
    xlab("") + 
    theme_bw() +
    ylab("median relative difference") +  
    theme(legend.position = "none",
          axis.text.y=element_text(size=16),
          axis.text.x=element_text(size=17),
          axis.title.y=element_text(size=20, vjust=1),
          axis.title.x=element_text(size=20) ) +
      ylim(0, 1)
```

```{r}
which_paralogs <- mr_paralog[[1]]$m_est_counts %>%
  dplyr::select(target_id) %>%
  distinct() %>%
  mutate(paralog = TRUE)

counts_dist <- mr[[1]]$m_est_counts %>%
  dplyr::select(target_id, oracle) %>%
  distinct() %>%
  left_join(which_paralogs, by = c("target_id"))

counts_dist <- counts_dist %>%
  mutate(paralog = ifelse(is.na(paralog), FALSE, TRUE)) %>%
  mutate(lab = ifelse(paralog, "Duplicated Genes Database", "Remaining transcripts")) %>%
  mutate(lab = factor(lab))
```

```{r}
ggplot(counts_dist, aes(oracle)) +
  geom_histogram(aes(y = ..density..)) +
  scale_x_log10() +
  facet_wrap(~ lab) +
  theme_bw(20) +
  xlab("true counts")
```
